staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go

```diff
@@ -76,7 +79,7 @@ func DefaultStacktracePred(status int) bool {
 }
 
 // WithLogging wraps the handler with logging.
-func WithLogging(handler http.Handler, pred StacktracePred) http.Handler {
+func WithLogging(handler http.Handler, pred StacktracePred, observers ...ResponseObserver) http.Handler {
```

Comment:
making the logging wrapper a general purpose hook for injecting callbacks to other things makes it a lot harder to reason about this filter... how does the API server record status code metrics?

Reply:
see the line 948-955 of  pkg/kubelet/server/server.go in this MR.

explain why we modify here:

in kubelet, it use  httplog.WithLogging to get handler ,like this
https://github.com/kubernetes/kubernetes/blob/77866160efce1ce4c3885d436b1cebaedc80c89c/pkg/kubelet/server/server.go#L930

and serve here:

https://github.com/kubernetes/kubernetes/blob/77866160efce1ce4c3885d436b1cebaedc80c89c/pkg/kubelet/server/server.go#L952

however, in handler, it use context to store  the response status, like this:
https://github.com/kubernetes/kubernetes/blob/77866160efce1ce4c3885d436b1cebaedc80c89c/staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go#L86

So from the outside (caller), we can't get response status from `req`.

Reply:
Injecting callbacks intended to be invoked even when we aren't logging the request makes this pretty hard to understand. How does the regular API server handle getting the status code to record metrics?